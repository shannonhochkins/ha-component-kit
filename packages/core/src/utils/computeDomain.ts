import { AllDomains, CamelToSnake } from "@typings";
import { snakeCase } from "lodash";

export const computeDomain = <E extends `${AllDomains}.${string}` | "unknown">(
  entityId: E
): CamelToSnake<AllDomains> =>
  snakeCase(
    entityId.substr(0, entityId.indexOf("."))
  ) as CamelToSnake<AllDomains>;
