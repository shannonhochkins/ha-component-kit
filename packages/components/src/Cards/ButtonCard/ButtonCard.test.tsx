// important that these are imported first or the mock won't work.
import {
  TestWrapper,
  onReady,
  mocked,
  connection,
} from "@mocks/mockConnection";
import { ButtonCard } from "@components";
import { computeDomain } from "@utils/computeDomain";
import { render, waitFor, fireEvent } from "@testing-library/react";
import { snakeCase } from "lodash";
describe("<ButtonCard />", () => {
  beforeEach(() => {
    onReady.mockClear();
  });
  // Test to check if the component renders correctly
  it("renders ButtonCard", async () => {
    const { getByTestId } = render(
      <ButtonCard
        entity="light.fake_light"
        service="turnOn"
        data-testid="button-card"
      />,
      {
        wrapper: TestWrapper,
      }
    );
    await waitFor(() => expect(onReady).toHaveBeenCalledTimes(1));
    const buttonElement = getByTestId("button-card");
    expect(buttonElement).toBeInTheDocument();
  });

  // Tests for all valid combinations of entity and service props
  const entities = ["light.fake_light", "switch.fake_gaming_switch"] as const;
  const services = ["turnOn", "toggle", "turnOff"] as const;

  entities.forEach((entity) => {
    services.forEach((service) => {
      it(`renders correctly with entity ${entity} and service ${service}`, async () => {
        const { getByTestId } = render(
          <ButtonCard
            entity={entity}
            service={service}
            data-testid="button-card"
          />,
          {
            wrapper: TestWrapper,
          }
        );
        await waitFor(() => expect(onReady).toHaveBeenCalledTimes(1));
        const buttonElement = getByTestId("button-card");
        expect(buttonElement).toBeInTheDocument();
      });

      it(`triggers service ${service} when clicked with entity ${entity}`, async () => {
        const mockFunction = jest.fn();
        const { getByTestId } = render(
          <ButtonCard
            entity={entity}
            service={service}
            onClick={mockFunction}
            data-testid="button-card"
          />,
          {
            wrapper: TestWrapper,
          }
        );
        await waitFor(() => expect(onReady).toHaveBeenCalledTimes(1));
        const buttonElement = getByTestId("button-card");
        fireEvent.click(buttonElement);
        expect(mocked.callService).toHaveBeenLastCalledWith(
          connection,
          computeDomain(entity),
          snakeCase(service),
          undefined,
          {
            entity_id: entity,
          }
        );
        expect(mockFunction).toHaveBeenCalled();
      });
    });
  });

  it("should render without an entity", async () => {
    const { getByTestId } = render(
      <ButtonCard
        title="something"
        description="somethingelse"
        icon="mdi:cross"
        data-testid="button-card"
      />,
      {
        wrapper: TestWrapper,
      }
    );
    await waitFor(() => expect(onReady).toHaveBeenCalledTimes(1));
    const buttonElement = getByTestId("button-card");
    expect(buttonElement).toBeInTheDocument();
    expect(buttonElement).toMatchSnapshot();
  });
});
