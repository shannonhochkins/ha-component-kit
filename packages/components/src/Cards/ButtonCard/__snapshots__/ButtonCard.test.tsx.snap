// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonCard /> should render without an entity 1`] = `
.emotion-0 {
  all: unset;
  padding: 1rem;
  position: relative;
  overflow: hidden;
  border-radius: 1rem;
  width: var(--ha-device-button-card-width);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  cursor: pointer;
  background-color: var(--ha-primary-background);
  box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.1);
  -webkit-transition: var(--ha-transition-duration) var(--ha-easing);
  transition: var(--ha-transition-duration) var(--ha-easing);
  transition-property: background-color,box-shadow,transform;
}

.emotion-0:active {
  -webkit-transform: translateY(5px) scale(0.98);
  -moz-transform: translateY(5px) scale(0.98);
  -ms-transform: translateY(5px) scale(0.98);
  transform: translateY(5px) scale(0.98);
  box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.1);
}

.emotion-0:hover {
  background-color: var(--ha-primary-background-hover);
  box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.1);
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  margin-bottom: 20px;
  gap: 10px;
}

.emotion-2 {
  border-radius: 100%;
  padding: 6px;
  width: 24px;
  height: 24px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  background-color: var(--ha-primary-inactive);
  color: var(--ha-secondary-inactive);
  -webkit-filter: brightness(100%);
  filter: brightness(100%);
  -webkit-transition: var(--ha-transition-duration) var(--ha-easing);
  transition: var(--ha-transition-duration) var(--ha-easing);
  transition-property: background-color,color,filter;
}

.emotion-2 svg {
  -webkit-transition: var(--ha-transition-duration) var(--ha-easing);
  transition: var(--ha-transition-duration) var(--ha-easing);
  transition-property: color;
}

.emotion-3 {
  position: relative;
  background-color: var(--ha-primary-inactive);
  border-radius: 10px;
  width: 40px;
  height: 20px;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: background-color var(--ha-transition-duration) var(--ha-easing);
  transition: background-color var(--ha-transition-duration) var(--ha-easing);
  margin-left: 20px;
}

.emotion-4 {
  background-color: white;
  border-radius: 100%;
  width: 16px;
  height: 16px;
  position: absolute;
  top: 2px;
  left: 0;
  -webkit-transition: var(--ha-transition-duration) var(--ha-easing);
  transition: var(--ha-transition-duration) var(--ha-easing);
  transition-property: left,transform;
  left: 0px;
  -webkit-transform: translate3d(calc(0% + 2px), 0, 0);
  -moz-transform: translate3d(calc(0% + 2px), 0, 0);
  -ms-transform: translate3d(calc(0% + 2px), 0, 0);
  transform: translate3d(calc(0% + 2px), 0, 0);
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.emotion-6 {
  color: var(--ha-secondary-color);
  font-size: 0.7rem;
  margin: 2px 0;
}

.emotion-8 {
  color: var(--ha-primary-color);
  font-size: 0.8rem;
}

<button
  class="emotion-0"
  data-testid="button-card"
>
  <div
    class="emotion-1"
  >
    <div
      class="emotion-2"
    >
      <span />
    </div>
    <div
      class="emotion-3"
    >
      <div
        class="emotion-4"
      />
    </div>
  </div>
  <div
    class="emotion-5"
  >
    <div
      class="emotion-6"
    >
      <div
        class="emotion-6"
      >
        something
      </div>
      <div
        class="emotion-8"
      >
        somethingelse
      </div>
    </div>
  </div>
</button>
`;
